File: apps/frontend/src/state/claudeCode.ts
import { create } from 'zustand';
import type { UnlistenFn } from '@tauri-apps/api/event';
import {
  claudeAvailable,
  claudePermissionRespond,
  claudeTurnCancel,
  claudeTurnStart,
  isTauri,
  onClaudeTurn,
  type ClaudeStreamEvent,
} from '../lib/tauri';
import { useFiles } from './files';
import { useSettings } from './settings';
import { useWorkspace } from './workspace';

/**
 * Claude Code integration state.
 *
 * The shape deliberately mirrors `state/wingman.ts` — same transcript rows,
 * same event vocabulary — because the two panels do the same job and ARC
 * shouldn't grow two mental models for "an agent said something". What differs
 * is underneath: there is no daemon and no connection. ARC spawns the user's
 * own `claude` binary per turn, so the only gate is whether it's installed.
 *
 * Conversation history lives in the CLI, not here: each turn reports a
 * `sessionId`, and passing it back as `resume` continues the same thread.
 */

/** Feature gate. `unavailable` means the CLI isn't on PATH — the common case
 *  for users who don't have it, and never an error. */
export type ClaudeStatus = 'checking' | 'ready' | 'unavailable';

/** One rendered transcript row. Identical vocabulary to the Wingman panel's. */
export type ClaudeChatItem =
  | { kind: 'user'; text: string }
  | { kind: 'assistant'; text: string }
  | { kind: 'thinking'; text: string }
  | { kind: 'tool'; id: string; name: string; input: unknown; output?: string; isError?: boolean }
  | { kind: 'error'; message: string }
  /** A permission the user answered. Kept in the transcript so the record of
   *  what was allowed sits next to what it then did. */
  | { kind: 'decision'; tool: string; allowed: boolean; summary: string };

/** A tool call the CLI is waiting on. The turn is genuinely paused while this
 *  is set — Claude is blocked on the answer, not merely slow. */
export interface ClaudePermission {
  requestId: string;
  tool: string;
  input: unknown;
  /** MCP tools can supply their own display block; plain tools have neither. */
  title: string | null;
  description: string | null;
  /** One-line rendering of what will actually run. */
  summary: string;
}

export interface ClaudeUsage {
  input: number;
  output: number;
  cacheRead: number;
  cacheWrite: number;
}

/** A file Claude wrote during this conversation, in the order first touched.
 *  This is the review queue: Claude Code edits the working tree directly, so
 *  the diff is already there — all ARC has to remember is which paths to open. */
export interface ClaudeEditedFile {
  /** Absolute path, as the tool call gave it. */
  path: string;
  /** Which tool wrote it — `Write` is a whole-file replacement, `Edit` a patch. */
  tool: string;
  /** Bumped when a later turn touches the same file again. */
  edits: number;
}

interface ClaudeState {
  status: ClaudeStatus;
  /** Resolved binary path — shown in Settings so "not installed" is diagnosable. */
  binary: string | null;

  /** CLI-held conversation id, from the turn's `init`. Null until the first turn. */
  sessionId: string | null;
  chat: ClaudeChatItem[];
  streaming: boolean;
  /** Topic of the in-flight turn, so it can be cancelled. */
  activeTopic: string | null;

  usage: ClaudeUsage | null;
  /** Cumulative USD across this conversation. The CLI bills per turn; the
   *  running total is the number that actually answers "what has this cost me". */
  costUsd: number;
  /** Tools Claude asked for and the permission mode refused, from the last
   *  turn's result. Surfaced because a denial otherwise looks like Claude
   *  simply choosing not to do the work. */
  denials: string[];

  editedFiles: ClaudeEditedFile[];

  /** The permission prompt currently blocking the turn, if any.
   *
   *  Only ever one: the CLI asks for a tool, waits, then asks for the next. A
   *  queue would imply a concurrency the protocol doesn't have. */
  pending: ClaudePermission | null;

  /** Probe for the CLI. Safe to call repeatedly; called on boot. */
  detect: () => Promise<void>;
  /** Answer the pending permission prompt and unblock the turn. */
  respond: (allow: boolean, message?: string) => Promise<void>;
  send: (prompt: string) => Promise<void>;
  /** Stop the in-flight turn. */
  cancel: () => Promise<void>;
  /** Drop the conversation and start a fresh session on the next turn. */
  newChat: () => void;
  /** Open one of Claude's edits in ARC's diff viewer, against the working tree. */
  openEditedFile: (path: string) => void;
}

/** Teardown handle for the live turn. Outside the store so it never lands in a
 *  React render path or a devtools snapshot. */
let unlistenTurn: UnlistenFn | null = null;

function num(v: unknown): number {
  return typeof v === 'number' && Number.isFinite(v) ? v : 0;
}

function str(v: unknown): string {
  return typeof v === 'string' ? v : '';
}

/** Tools whose input names a file Claude is about to write.
 *
 *  Read-only tools are deliberately absent: the review list exists to answer
 *  "what changed", and padding it with every file Claude looked at would bury
 *  the two it edited. */
const WRITE_TOOLS: Record<string, string> = {
  Edit: 'file_path',
  Write: 'file_path',
  MultiEdit: 'file_path',
  NotebookEdit: 'notebook_path',
};

/** Pull the written path out of a tool call, or null if it wrote nothing. */
export function editedPath(name: string, input: unknown): string | null {
  const key = WRITE_TOOLS[name];
  if (!key || !input || typeof input !== 'object') return null;
  const path = (input as Record<string, unknown>)[key];
  return typeof path === 'string' && path.trim() ? path : null;
}

export const useClaudeCode = create<ClaudeState>((set, get) => ({
  status: 'checking',
  binary: null,
  sessionId: null,
  chat: [],
  streaming: false,
  activeTopic: null,
  usage: null,
  costUsd: 0,
  denials: [],
  editedFiles: [],
  pending: null,

  detect: async () => {
    // Browser-only dev build has no IPC. Report unavailable rather than
    // throwing — the panel's disconnected state is already the right UI.
    if (!isTauri) {
      set({ status: 'unavailable', binary: null });
      return;
    }
    try {
      const binary = await claudeAvailable();
      set({ status: binary ? 'ready' : 'unavailable', binary });
    } catch {
      set({ status: 'unavailable', binary: null });
    }
  },

  newChat: () => {
    unlistenTurn?.();
    unlistenTurn = null;
    set({
      sessionId: null,
      chat: [],
      streaming: false,
      activeTopic: null,
      usage: null,
      costUsd: 0,
      denials: [],
      editedFiles: [],
      pending: null,
    });
  },

  respond: async (allow, message) => {
    const { pending, activeTopic } = get();
    if (!pending || !activeTopic) return;
    // Clear before the call, not after: the CLI resumes the moment it reads
    // the answer, and a prompt still on screen would invite a second click
    // that lands on a request id the turn has already moved past.
    set((s) => ({
      pending: null,
      chat: [
        ...s.chat,
        { kind: 'decision', tool: pending.tool, allowed: allow, summary: pending.summary },
      ],
    }));
    try {
      await claudePermissionRespond({
        topic: activeTopic,
        requestId: pending.requestId,
        allow,
        message: message ?? null,
      });
    } catch (e) {
      set((s) => ({
        chat: [
          ...s.chat,
          { kind: 'error', message: e instanceof Error ? e.message : String(e) },
        ],
      }));
    }
  },

  cancel: async () => {
    const topic = get().activeTopic;
    if (!topic) return;
    // The turn's own terminal event flips `streaming` — killing the child
    // produces one, so there's nothing to reset here.
    await claudeTurnCancel(topic).catch(() => {});
  },

  openEditedFile: (path) => {
    // Claude Code edits the working tree in place, so the diff against HEAD is
    // exactly what ARC's Source Control already renders. Root at the workspace,
    // not at the file — that's what the git stack expects.
    const root = useFiles.getState().root;
    if (!root) return;
    useWorkspace.getState().openDiff(path, root, 'worktree');
  },

  send: async (prompt) => {
    const { status, streaming } = get();
    const root = useFiles.getState().root;
    // One turn at a time: a second child would be a second conversation, and
    // both would race to resume the same session id.
    if (status !== 'ready' || streaming || !prompt.trim()) return;
    if (!root) {
      set((s) => ({
        chat: [
          ...s.chat,
          { kind: 'error', message: 'Open a folder before asking Claude about it.' },
        ],
      }));
      return;
    }

    const settings = useSettings.getState();
    set((s) => ({
      chat: [...s.chat, { kind: 'user', text: prompt }],
      streaming: true,
      denials: [],
      pending: null,
    }));

    try {
      const topic = await claudeTurnStart({
        cwd: root,
        prompt,
        resume: get().sessionId,
        model: settings.claudeModel || null,
        permissionMode: settings.claudePermissionMode,
        maxBudgetUsd: settings.claudeMaxBudgetUsd || null,
      });
      set({ activeTopic: topic });

      unlistenTurn?.();
      unlistenTurn = await onClaudeTurn(topic, (ev) => applyEvent(set, ev));
    } catch (e) {
      set((s) => ({
        chat: [
          ...s.chat,
          { kind: 'error', message: e instanceof Error ? e.message : String(e) },
        ],
        streaming: false,
        activeTopic: null,
      }));
    }
  },
}));

/**
 * Fold one stream event into the transcript.
 *
 * Text and thinking deltas append to the trailing row of the same kind rather
 * than pushing a new one — otherwise a turn produces one row per token.
 */
function applyEvent(
  set: (fn: (s: ClaudeState) => Partial<ClaudeState>) => void,
  ev: ClaudeStreamEvent,
): void {
  const p = ev.payload ?? {};

  switch (ev.kind) {
    // The CLI mints the session id, so ARC learns it from the first turn and
    // replays it as `resume` from then on.
    case 'init':
      set(() => ({ sessionId: str(p.session_id) || null }));
      return;

    case 'text_delta':
    case 'thinking_delta': {
      const kind = ev.kind === 'text_delta' ? 'assistant' : 'thinking';
      const text = str(p.text);
      if (!text) return;
      set((s) => {
        const last = s.chat[s.chat.length - 1];
        if (last && last.kind === kind) {
          const chat = s.chat.slice(0, -1);
          chat.push({ kind, text: last.text + text } as ClaudeChatItem);
          return { chat };
        }
        return { chat: [...s.chat, { kind, text } as ClaudeChatItem] };
      });
      return;
    }

    // The turn is now blocked on the user. Nothing else arrives until
    // `respond` answers it or the turn is cancelled.
    case 'permission_request': {
      const tool = str(p.display_name) || str(p.tool_name) || 'tool';
      set(() => ({
        pending: {
          requestId: str(p.request_id),
          tool,
          input: p.input,
          title: str(p.title) || null,
          description: str(p.description) || null,
          summary: summarizeToolInput(str(p.tool_name), p.input),
        },
      }));
      return;
    }

    case 'tool_start': {
      const name = str(p.name);
      const path = editedPath(name, p.input);
      set((s) => ({
        chat: [...s.chat, { kind: 'tool', id: str(p.id), name, input: p.input }],
        // Record the edit at call time, not on the result: a tool that fails
        // still leaves the file worth looking at, and the reviewer is better
        // served by a diff that turns out empty than by a missing entry.
        editedFiles: path ? withEdit(s.editedFiles, path, name) : s.editedFiles,
      }));
      return;
    }

    case 'tool_result':
      // Match the result onto its call so the UI shows one row per tool, not a
      // call row and an orphaned result row.
      set((s) => {
        const id = str(p.id);
        const idx = s.chat.findIndex((c) => c.kind === 'tool' && c.id === id);
        if (idx === -1) return {};
        const chat = [...s.chat];
        const row = chat[idx] as Extract<ClaudeChatItem, { kind: 'tool' }>;
        chat[idx] = { ...row, output: str(p.output), isError: Boolean(p.is_error) };
        return { chat };
      });
      return;

    case 'usage': {
      const u = (p.usage ?? {}) as Record<string, unknown>;
      set(() => ({
        usage: {
          input: num(u.input_tokens),
          output: num(u.output_tokens),
          cacheWrite: num(u.cache_creation_input_tokens),
          cacheRead: num(u.cache_read_input_tokens),
        },
      }));
      return;
    }

    // The CLI's terminal line for a turn. `done` still follows (the child has
    // to exit), so this only records the turn's outcome and leaves `streaming`
    // to the process teardown.
    case 'result': {
      const denials = Array.isArray(p.permission_denials)
        ? p.permission_denials
            .map((d) =>
              typeof d === 'object' && d !== null
                ? str((d as Record<string, unknown>).tool_name)
                : str(d),
            )
            .filter(Boolean)
        : [];
      set((s) => ({
        sessionId: str(p.session_id) || s.sessionId,
        costUsd: s.costUsd + num(p.cost_usd),
        denials,
        // A turn can fail after producing output (budget exhausted, API error
        // mid-stream). The row makes that visible instead of leaving a reply
        // that just stops.
        chat:
          p.is_error === true
            ? [...s.chat, { kind: 'error', message: str(p.text) || 'the turn failed' }]
            : s.chat,
      }));
      return;
    }

    // Both terminal paths drop `pending`: a cancelled turn leaves a prompt
    // on screen that answers into a dead process otherwise.
    case 'error':
      set((s) => ({
        chat: [...s.chat, { kind: 'error', message: str(p.message) || 'the turn failed' }],
        streaming: false,
        activeTopic: null,
        pending: null,
      }));
      return;

    case 'done':
      set(() => ({ streaming: false, activeTopic: null, pending: null }));
      return;

    default:
      return;
  }
}

/**
 * One line describing what a tool call will actually do.
 *
 * This is the whole value of an approval prompt: "allow Bash?" is not a
 * decision anyone can make, and `rm -rf build` is. Each known tool has one
 * field that carries the substance; everything else falls back to compact JSON
 * rather than showing nothing.
 */
export function summarizeToolInput(tool: string, input: unknown): string {
  if (!input || typeof input !== 'object') return '';
  const o = input as Record<string, unknown>;
  const first = (...keys: string[]): string => {
    for (const k of keys) {
      const v = o[k];
      if (typeof v === 'string' && v.trim()) return v;
    }
    return '';
  };
  switch (tool) {
    case 'Bash':
    case 'BashOutput':
      return first('command');
    case 'Read':
    case 'Edit':
    case 'Write':
    case 'MultiEdit':
      return first('file_path');
    case 'NotebookEdit':
      return first('notebook_path');
    case 'Glob':
    case 'Grep':
      return first('pattern');
    case 'WebFetch':
    case 'WebSearch':
      return first('url', 'query');
    default: {
      const json = JSON.stringify(o);
      return json.length > 300 ? `${json.slice(0, 300)}…` : json;
    }
  }
}

/** Add or bump one entry in the edited-file list, preserving first-touch order. */
function withEdit(
  files: ClaudeEditedFile[],
  path: string,
  tool: string,
): ClaudeEditedFile[] {
  const at = files.findIndex((f) => f.path === path);
  const prev = at === -1 ? undefined : files[at];
  if (!prev) return [...files, { path, tool, edits: 1 }];
  const next = [...files];
  next[at] = { ...prev, tool, edits: prev.edits + 1 };
  return next;
}

/** Reset for tests. Not used by the app. */
export function __resetClaudeForTests(): void {
  unlistenTurn?.();
  unlistenTurn = null;
  useClaudeCode.setState({
    status: 'ready',
    sessionId: null,
    chat: [],
    streaming: false,
    activeTopic: null,
    usage: null,
    costUsd: 0,
    denials: [],
    editedFiles: [],
    pending: null,
  });
}

export { applyEvent as __applyEventForTests };

File: apps/desktop/src/commands/claude_code.rs
//! Tauri command surface for the Claude Code integration.
//!
//! Where Wingman is a daemon ARC connects to, Claude Code is a binary ARC
//! spawns — one child process per turn, reading its `stream-json` output (see
//! `arc-claude-code`). There is nothing to configure and nothing to connect:
//! if `claude` is on PATH the panel works, and if it isn't the frontend hides
//! the surface. Nothing else in ARC depends on this module.
//!
//! Frontend contract (see apps/frontend/src/lib/tauri.ts):
//!   invoke("claude_available")                          -> string | null
//!   invoke("claude_turn_start", { ... })                -> topic
//!   invoke("claude_turn_cancel", { topic })             -> ()
//!   invoke("claude_permission_respond", { ... })        -> ()
//!
//! `claude_turn_start` returns immediately with a topic and emits
//! `{ kind, payload }` on it per event, then exactly one terminal event —
//! `done` on a clean end, `error` when the child could not run at all.
//!
//! A turn can pause on a `permission_request` event and stay paused until
//! `claude_permission_respond` answers it, so the two commands are halves of
//! one conversation rather than independent calls.

use arc_claude_code::{run_turn, Decision, TurnOptions};
use dashmap::DashMap;
use std::sync::atomic::{AtomicU64, Ordering};
use tauri::{AppHandle, Emitter, Manager, State};
use tokio::sync::{mpsc, oneshot};

/// Per-turn back-channels, keyed by topic.
#[derive(Default)]
pub struct ClaudeState {
    /// Kill switch behind the panel's stop button.
    cancels: DashMap<String, oneshot::Sender<()>>,
    /// Where approve/deny answers go. Separate from `cancels` because a turn
    /// can take many decisions and exactly one cancellation.
    decisions: DashMap<String, mpsc::UnboundedSender<Decision>>,
    next_stream: AtomicU64,
}

impl ClaudeState {
    /// Drop both channels for a finished turn. Idempotent — a cancelled turn
    /// reaches this by two paths.
    fn forget(&self, topic: &str) {
        self.cancels.remove(topic);
        self.decisions.remove(topic);
    }
}

/// Resolve the Claude Code binary. `None` means "not installed" — the gate for
/// the whole feature, and the one command the frontend calls unconditionally.
#[tauri::command]
pub async fn claude_available() -> Result<Option<String>, String> {
    Ok(arc_claude_code::binary())
}

/// Start a turn and stream it.
///
/// `resume` continues a prior conversation by its `session_id`; omitting it
/// starts a fresh one. The child runs in `cwd`, which is what bounds Claude's
/// file access, so the frontend passes the workspace root rather than letting
/// the CLI infer one.
#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn claude_turn_start(
    app: AppHandle,
    state: State<'_, ClaudeState>,
    cwd: String,
    prompt: String,
    resume: Option<String>,
    model: Option<String>,
    permission_mode: Option<String>,
    max_budget_usd: Option<f64>,
) -> Result<String, String> {
    if cwd.trim().is_empty() {
        return Err("claude: no workspace folder is open".into());
    }

    let id = state.next_stream.fetch_add(1, Ordering::Relaxed);
    let topic = format!("claude://turn/{id}");

    let (cancel_tx, cancel_rx) = oneshot::channel();
    let (decision_tx, decision_rx) = mpsc::unbounded_channel();
    state.cancels.insert(topic.clone(), cancel_tx);
    state.decisions.insert(topic.clone(), decision_tx);

    let opts = TurnOptions {
        cwd,
        prompt,
        resume,
        model,
        permission_mode,
        max_budget_usd,
        binary: None,
    };

    let emit_topic = topic.clone();
    tauri::async_runtime::spawn(async move {
        let app2 = app.clone();
        let t2 = emit_topic.clone();
        let result = run_turn(opts, cancel_rx, decision_rx, move |ev| {
            let _ = app2.emit(
                &t2,
                serde_json::json!({ "kind": ev.kind, "payload": ev.payload }),
            );
        })
        .await;

        // A guaranteed last event: a child that fails to spawn produces no
        // stream at all, and without this the composer would sit on a spinner.
        let terminal = match result {
            Ok(()) => serde_json::json!({ "kind": "done", "payload": {} }),
            Err(e) => {
                tracing::warn!("claude turn ended: {e:#}");
                serde_json::json!({ "kind": "error", "payload": { "message": format!("{e:#}") } })
            }
        };
        let _ = app.emit(&emit_topic, terminal);

        if let Some(state) = app.try_state::<ClaudeState>() {
            state.forget(&emit_topic);
        }
    });

    Ok(topic)
}

/// Answer a `permission_request`. `message` is shown to Claude on a denial so
/// it can adapt instead of retrying the same call.
///
/// Unknown topics are a no-op: the turn may have been cancelled between the
/// prompt appearing and the user answering it.
#[tauri::command]
pub async fn claude_permission_respond(
    state: State<'_, ClaudeState>,
    topic: String,
    request_id: String,
    allow: bool,
    message: Option<String>,
) -> Result<(), String> {
    if let Some(tx) = state.decisions.get(&topic) {
        let _ = tx.send(Decision {
            request_id,
            allow,
            message: message.filter(|m| !m.trim().is_empty()),
        });
    }
    Ok(())
}

/// Kill a running turn. Unknown topics are a no-op — the turn may have just
/// finished on its own. This is also the escape hatch from a permission prompt
/// the user would rather not answer either way.
#[tauri::command]
pub async fn claude_turn_cancel(
    state: State<'_, ClaudeState>,
    topic: String,
) -> Result<(), String> {
    if let Some((_, tx)) = state.cancels.remove(&topic) {
        let _ = tx.send(());
    }
    Ok(())
}
